### ISSUE TITLE
Type-safe helpers for Durable Functions

### ISSUE THREAD

**Issue by stuartleeks** [Azure/azure-functions-core-tools#946]
## PROPOSAL 
Add a command to the CLI to generate type-safe helpers for working with Durable Functions in C#

(NOTE: it would be interesting to explore this for other languages as well - e.g. can we create TypeScript definitions to augment the JavaScript experience etc)

## BACKGROUND:
When working with Durable Functions in C# it would be nice to have a way to call activity functions (and sub-orchestrations) without using magic strings.

In the standard template there is a `SayHello` function:

```csharp
[FunctionName("Function1_Hello")]
public static string SayHello([ActivityTrigger] string name, TraceWriter log)
{
    log.Info($"Saying hello to {name}.");
    return $"Hello {name}!";
}
```

which is invoked using this code:

```csharp
outputs.Add(await context.CallActivityAsync<string>("Function1_Hello", "Tokyo"));
outputs.Add(await context.CallActivityAsync<string>("Function1_Hello", "Seattle"));
outputs.Add(await context.CallActivityAsync<string>("Function1_Hello", "London"));
```

When using the `CallActivityAsync` function the first parameter is a string identifying the function to invoke (based on the name in the `FunctionName` attribute on the function method) and the second parameter is the input to the function (which is typed as object in the method declaration). As a result the editor and compiler are unable to give any assistance to either parameter.

The proposal is to add a command to the CLI to generate type-safe helpers for activities (and orchestrations) so that both the editor and compiler are able to provide assistance. With these helpers the orchestration code becomes:

```csharp
outputs.Add(await context.Activities().SayHelloAsync("Tokyo"));
outputs.Add(await context.Activities().SayHelloAsync("Seattle"));
outputs.Add(await context.Activities().SayHelloAsync("London"));
```

Key features:
 - use method name as the name of the generated method as this feels natural in the code
 - use `Activities()` top-level grouping. This will help with navigating the methods when the number of activity functions grows, as well as helping to reduce the chances of name clashes with methods on the context
 - strongly type the parameter to the generated method, and re-use the name from parameter in the method that implements the activity function to give added context

**Comment by stimms** [comment]
First off I love this suggestion. I just fixed a bug in one of our durable functions which was due to types being incorrectly passed into a child function. I'm not sure if a command line tool is the best approach for a few reasons:

1. Discoverability - Unless you know to run the command line generator you wouldn't get benefit from this
2. Synchronization - If you make a signature change to a function and don't regenerate the helpers then you will still be under the belief that everything will work when it actually won't.

I was thinking about a T3 template but it has the same limitations as above. What I was thinking was maybe a Rosslyn analyser which would do the checking. You'd probably lose the nice intellisense completions you get with generated stubs but it would solve 1 and 2.

**Comment by stuartleeks** [comment]
Thanks @stimms - glad it helped you find a bug!

My [initial explorations](https://github.com/stuartleeks/DurableFunctionsTypedSafeActivities) for this idea were actually implemented in a T4 template. Unfortunately it required you to be running in Visual Studio to execute the template.

My aim with converting to a CLI command is that it can run cross platform and easily be brought into a build pipeline.

I'd also like to explore automatically running this command when the project is built, ideally having that change included in the templates :-)

I hadn't thought about Roslyn analysers to give you a more immediate prompt if your calls aren't valid, but that would be interesting to explore.

**Comment by stuartleeks** [comment]
Thinking about the Roslyn analysers, I wonder if you could use them to detect a change to an activity signature that isn't reflected in the generated type-safe helpers? With a codefix action to re-run the generation!

**Comment by amadard** [comment]
Are there any updates on this item?  I am searching around for other options to solve these exact problems with FunctionName, input and output in C# code.  I can manually create the Activities() setup (Thanks for that!), but if it was an feature of the DurableFunctions it would be very helpful.

**Comment by stimms** [comment]
I made some analyzers https://blog.simontimms.com/2019/02/17/2019-02-17-durable_functions_analyzer/ but as far as I know nothing was added to the base product.

**Comment by amadard** [comment]
Thanks @stimms, I looked over your analyzers.  We are exclusively using .GetInput<> for all of our Activities and SubOrchestrators input, so I went the route of building my own ActivitiesManager class.  I added a custom Attribute for the inputs, so the class can be built through code.

Thanks @stuartleeks for your initial code examples, I was able to build off of those.

**Comment by ahmedelnably** [comment]
cc/ @anthonychu

**Comment by WhiteTomX** [comment]
Hey,
i hate to resurrect such an old issue but i just started with Durable Functions and wonder if it is really a good way to work with MagicStrings and hope that the parameter is correct. So maybe some of you found a good way yet. I looked around and found [Type-safe activity helper for Durable Functions](https://github.com/shibayan/durable-functions-typed-proxy). But didn't have the time to try yet.

**Comment by cgillum** [comment]
Thanks for pinging this thread. We decided to take a slightly different approach to this and leverage C# source generators. It automatically creates type-safe wrappers for activity calls in a way that doesn't require any specific tooling. It's available starting in the [v2.5.0](https://github.com/Azure/azure-functions-durable-extension/releases/tag/v2.5.0) release of the Durable Functions extension. More information about how it works and how to start leveraging it can be found [here](https://github.com/Azure/azure-functions-durable-extension/blob/dev/src/DurableFunctions.TypedInterfaces/README.md#codegeneration-for-durable-functions).
